autoprotocol/protocol.py

```diff
@@ -5135,10 +5141,31 @@ def provision(self, resource_id, dests, volumes):
                 self.instructions[-1].to.append(xfer)
             else:
                 provision_instructions_to_return.append(
-                    self._append_and_return(Provision(resource_id, dest_group))
+                    self._append_and_return(
+                        Provision(resource_id, dest_group, measurement_mode)
+                    )
                 )
         return provision_instructions_to_return
 
+    def _identify_measurement_mode(self, volumes):
+        unique_measure_modes = set()
+        for amount in volumes:
+            print(amount.dimensionality)
+            if not isinstance(amount, (str, Unit)):
+                raise TypeError(f"Provided amount {amount} is not supported.")
+            if amount.dimensionality == Unit(1, "liter").dimensionality:
+                unique_measure_modes.add("volume")
+            elif amount.dimensionality == Unit(1, "gram").dimensionality:
+                unique_measure_modes.add("mass")
+            else:
+                raise ValueError(
+                    f"Provisioning of resources with measurement unit of {amount.unit} is not supported."
+                )
+        if len(unique_measure_modes) != 1:
+            raise ValueError("Received amounts with more than one measurement mode")
+        measurement_mode = unique_measure_modes.pop()
```

Comment:
nit: would this be a good place to check that the single item in `unique_measure_mode` is also in `PROVISION_MEASUREMENT_MODES`?

Reply:
I did thought about it, but it will be redundant as we already have that check in the `Provision` constructor.

Reply:
Yeah, I was thinking that the check here would be in place of that one. I'll leave it up to you

Reply:
If someone is trying to create provision object like :
```
provision = Provision(resource_id, dest_group, measurement_mode)
```
the check in the constructor will prevent invalid measurement_mode being set in the object